### PR TITLE
subscriber: merge parent fields for json formatter

### PR DESCRIPTION
**Pull request by jeromegn**
Adds a `merge_parent_fields` option to the JSON formatter which propagates all fields from parent spans into the current event's fields when formatting.

## Motivation

It's easier to parse a flatter tree of JSON. We wanted to know about all fields from parent spans without using the `with_span_list(true)` option.

## Solution

- Add a `merge_parent_fields(enable: bool, namespace: bool)` option
- Walk up the span tree for each event and add their parents' fields to its own
- Namespaces (if set to `true`) each field with the parent's name and a `.`.

Example:

```rust
tracing_subscriber::fmt::Layer::new()
                        .json()
                        .with_span_list(false)
                        .with_current_span(false)
                        .merge_parent_fields(true, true);
```

```rust
let parent = tracing::span!(tracing::Level::INFO, "parent", is_parent = true);
let _parent_guard = parent.enter();
let span = tracing::span!(tracing::Level::INFO, "json_span", answer = 42, number = 3);
let _guard = span.enter();
tracing::info!(answer = 42, number = 4, "some json test");
```

This ends up looking like:

```json
{
    "timestamp": "fake time",
    "level": "INFO",
    "fields": {
        "message": "some json test",
        "answer": 42,
        "number": 4,
        "json_span.answer": 42,
        "json_span.number": 3,
        "parent.is_parent": true
    },
    "target": "tracing_subscriber::fmt::format::json::test"
}
```

### Caveats

- This is a clumsy and pretty slow implementation. It has to deseralize and serialize again every parent span's already-formatted fields.
- Setting the namespace option to false can lead to duplicate field names and that's handled different across the ecosystem.

Possible fix for these caveats would be to make a full `Layer` (I think they're now `Collect`?) that stores `serde_json::Value`s instead of strings.

I'm not entirely sure I will use these changes in our project because of the caveats, but I've been told they might be useful to other people, so made a PR.

### CODE REVIEW

**Comment by jeromegn** [comment]
I ended up making a `Layer` like this in our project:

```rust
use std::io::Write;
use std::{collections::BTreeMap, fmt};

use chrono::{SecondsFormat, Utc};
use serde::{ser::SerializeMap, Serializer};
use tracing::{
    span::{Attributes, Record},
    Event, Id, Subscriber,
};
use tracing_core::Field;
use tracing_serde::AsSerde;
use tracing_subscriber::{
    field,
    layer::{Context, Layer},
    registry::LookupSpan,
};

pub struct JsonLayer;

impl<S> Layer<S> for JsonLayer
where
    S: Subscriber + for<'a> LookupSpan<'a>,
{
    fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
        let span = ctx.span(id).expect("Span not found, this is a bug");
        let mut extensions = span.extensions_mut();

        if extensions.get_mut::<JsonFields>().is_none() {
            let mut json_fields = JsonFields { values: BTreeMap::new() };

            attrs.record(&mut json_fields);
            extensions.insert(json_fields);
        }
    }

    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
        let span = ctx.span(id).expect("Span not found, this is a bug");
        let mut extensions = span.extensions_mut();
        if let Some(json_fields) = extensions.get_mut::<JsonFields>() {
            values.record(json_fields);
        } else {
            let mut json_fields = JsonFields { values: BTreeMap::new() };

            values.record(&mut json_fields);
            extensions.insert(json_fields);
        }
    }

    fn on_enter(&self, _id: &Id, _ctx: Context<'_, S>) {}

    fn on_exit(&self, _id: &Id, _ctx: Context<'_, S>) {}

    fn on_close(&self, _id: Id, _ctx: Context<'_, S>) {}

    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
        let now = Utc::now();

        let mut event_fields = JsonFields { values: BTreeMap::new() };
        // event_fields.values.insert("message", serde_json::Value::from(event.))
        event.record(&mut event_fields);

        let current = ctx.current_span();
        self.merge_parent_fields(
            &mut event_fields.values,
            &ctx,
            event.parent().or_else(|| if event.is_contextual() { current.id() } else { None }),
        );

        let mut serializer = serde_json::Serializer::new(std::io::stdout());
        let mut serializer = serializer.serialize_map(None).expect("could not make json map serializer");

        let meta = event.metadata();

        serializer
            .serialize_entry("timestamp", &now.to_rfc3339_opts(SecondsFormat::Micros, true))
            .expect("could not serialize timestamp");
        serializer
            .serialize_entry("level", &meta.level().as_serde())
            .expect("could not serialize level");
        serializer.serialize_entry("fields", &event_fields.values).expect("could not serialize fields");
        serializer.serialize_entry("target", meta.target()).expect("could not serialize target");
        serializer.end().expect("could not end map");

        writeln!(std::io::stdout()).expect("could not write new line");
    }
}

impl JsonLayer {
    fn merge_parent_fields<S: Subscriber + for<'a> LookupSpan<'a>>(
        &self,
        btree: &mut BTreeMap<&str, serde_json::Value>,
        ctx: &Context<'_, S>,
        span_id: Option<&Id>,
    ) {
        if let Some(span_id) = span_id {
            let span = ctx.span(span_id).expect("no span, this is a bug");
            let ext = span.extensions();
            if let Some(json_fields) = ext.get::<JsonFields>() {
                for (k, v) in json_fields.values.iter() {
                    btree.entry(k).or_insert_with(|| v.clone());
                }
            }
            self.merge_parent_fields(btree, ctx, span.parent_id());
        }
    }
}

struct JsonFields<'a> {
    values: BTreeMap<&'a str, serde_json::Value>,
}

impl<'a> field::Visit for JsonFields<'a> {
    /// Visit a signed 64-bit integer value.
    fn record_i64(&mut self, field: &Field, value: i64) {
        self.values.insert(&field.name(), serde_json::Value::from(value));
    }

    /// Visit an unsigned 64-bit integer value.
    fn record_u64(&mut self, field: &Field, value: u64) {
        self.values.insert(&field.name(), serde_json::Value::from(value));
    }

    /// Visit a boolean value.
    fn record_bool(&mut self, field: &Field, value: bool) {
        self.values.insert(&field.name(), serde_json::Value::from(value));
    }

    /// Visit a string value.
    fn record_str(&mut self, field: &Field, value: &str) {
        self.values.insert(&field.name(), serde_json::Value::from(value));
    }

    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
        match field.name() {
            name if name.starts_with("r#") => {
                self.values.insert(&name[2..], serde_json::Value::from(format!("{:?}", value)));
            }
            name => {
                self.values.insert(name, serde_json::Value::from(format!("{:?}", value)));
            }
        };
    }
}
```

Helped a lot to have done this PR before :)

**Review by hawkw — COMMENTED** [review]
this looks like a great start! i think there are a few things that could be improved --- i left some comments. let me know what you think, especially about the API design?

File: tracing-subscriber/src/fmt/format/json.rs [hunk 1]
```diff
@@ -165,13 +198,106 @@ where
...
         serializer.serialize_entry("name", self.0.metadata().name())?;
         serializer.end()
     }
 }
 
+struct MergeParentFieldsMap<'a, S, N> {
+    event: &'a Event<'a>,
+    ctx: &'a FmtContext<'a, S, N>,
+    namespace: bool,
+    phantom: std::marker::PhantomData<N>,
```
[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/format/json.rs:229]: I'm not entirely sure! I barely understand PhantomData 😊. it's at least necessary to pass to `SerializableSpan` later on, unless that one also doesn't need it. I think it still does need it?

File: tracing-subscriber/src/fmt/format/json.rs [hunk 2]
```diff
@@ -165,13 +198,106 @@ where
...
+            map,
+            self.event.parent().or_else(|| {
+                if self.event.is_contextual() {
+                    current.id()
+                } else {
+                    None
+                }
+            }),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/format/json.rs:278]: That's a lot simpler! I love it.

File: tracing-subscriber/src/fmt/fmt_subscriber.rs [hunk 3]
```diff
@@ -424,6 +424,28 @@ impl<S, T, W> Subscriber<S, format::JsonFields, format::Format<format::Json, T>,
...
+    ///
+    /// `enable` argument enables or disables the option.
+    /// `namespace` argument sets whether or not fields should be namespaced with the
+    /// parent's span name
+    ///
+    /// See [`format::Json`]
+    pub fn merge_parent_fields(
+        self,
+        enable: bool,
+        namespace: bool,
```
[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/fmt_subscriber.rs]: Changed this to 2 fields as suggested!

File: tracing-subscriber/src/fmt/format/json.rs [hunk 4]
```diff
@@ -96,32 +108,49 @@ where
...
+
+        serializer.serialize_str(self.key)
+    }
+}
+
 struct SerializableSpan<'a, 'b, Span, N>(
     &'b crate::registry::SpanRef<'a, Span>,
+    Option<&'static str>,
     std::marker::PhantomData<N>,
 )
```
[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/format/json.rs]: Makes sense!

File: tracing-subscriber/src/fmt/fmt_subscriber.rs [hunk 5]
```diff
@@ -424,6 +424,28 @@ impl<S, T, W> Subscriber<S, format::JsonFields, format::Format<format::Json, T>,
...
+    ) -> Subscriber<S, format::JsonFields, format::Format<format::Json, T>, W> {
+        Subscriber {
+            fmt_event: self.fmt_event.merge_parent_fields(enable, namespace),
+            fmt_fields: format::JsonFields::new(),
+            fmt_span: self.fmt_span,
+            make_writer: self.make_writer,
+            _inner: self._inner,
```
[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/fmt_subscriber.rs]: TIL

File: tracing-subscriber/src/fmt/format/json.rs [hunk 6]
```diff
@@ -543,6 +690,62 @@ mod test {
         });
     }
 
+    #[test]
+    fn json_merged_parent_fields() {
+        let expected =
+        "{\"timestamp\":\"fake time\",\"level\":\"INFO\",\"fields\":{\"message\":\"some json test\",\"answer\":42,\"number\":4,\"json_span.answer\":42,\"json_span.number\":3,\"parent.is_parent\":true},\"target\":\"tracing_subscriber::fmt::format::json::test\"}\n";
```
[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/format/json.rs:693]: I'm adding a note saying the user is responsible for ensuring there are no duplicate fields in the tree of parent spans.

The alternative would be to track "seen" fields via a `HashSet` or something like that. I don't know if that's desirable? I don't mind adding that logic, but it's not great.

[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/format/json.rs:693]: The `json_merged_parent_fields_no_namespace` already tests this. It shows the same `answer` field twice.

File: tracing-subscriber/src/fmt/fmt_subscriber.rs [hunk 7]
```diff
@@ -424,6 +424,28 @@ impl<S, T, W> Subscriber<S, format::JsonFields, format::Format<format::Json, T>,
             ..self
         }
     }
+
+    /// Sets whether or not the formatter will merge events' parent spans fields
+    /// into the `fields` of this event
+    ///
+    /// `enable` argument enables or disables the option.
+    /// `namespace` argument sets whether or not fields should be namespaced with the
```
[the remark this answers is not in the document]
Comment (jeromegn) [reply tracing-subscriber/src/fmt/fmt_subscriber.rs]: Yes, added a note.

**Comment by jeromegn** [comment]
I've made the progress I could make without further feedback. Happy to make more changes!

Somewhat related: I've changed our `Layer` implementation so it doesn't have clone every `serde_json::Value`.

```rust
struct FieldsSerializer<'a, S> {
    ctx: &'a Context<'a, S>,
    fields: JsonFields<'a>,
}

impl<'a, S> serde::Serialize for FieldsSerializer<'a, S>
where
    S: Subscriber + for<'b> LookupSpan<'b>,
{
    fn serialize<Ser>(&self, serializer: Ser) -> Result<Ser::Ok, Ser::Error>
    where
        Ser: serde::Serializer,
    {
        let mut map = serializer.serialize_map(None)?;
        let mut seen: HashSet<&str> = HashSet::new();
        for (k, v) in self.fields.values.iter() {
            map.serialize_entry(k, v)?;
            seen.insert(k);
        }

        if let Some(span) = self.ctx.lookup_current() {
            let ext = span.extensions();
            if let Some(json_fields) = ext.get::<JsonFields>() {
                for (k, v) in json_fields.values.iter() {
                    if !seen.contains(k) {
                        map.serialize_entry(k, v)?;
                        seen.insert(k);
                    }
                }
            }

            for span in span.parents() {
                let ext = span.extensions();
                if let Some(json_fields) = ext.get::<JsonFields>() {
                    for (k, v) in json_fields.values.iter() {
                        if !seen.contains(k) {
                            map.serialize_entry(k, v)?;
                            seen.insert(k);
                        }
                    }
                }
            }
        }

        map.end()
    }
}
```

**Comment by jeromegn** [comment]
Any update on this? 2 outstanding comments I'm not sure what to do about :)

**Comment by hawkw** [comment]
> Any update on this? 2 outstanding comments I'm not sure what to do about :)

Whoops, sorry I missed this --- lemme try to give this branch another review so we can get it merged!

**Comment by bryangarza** [comment]
AFAICT, with a rebased and another round of review, this could still be merged

**Comment by jeromegn** [comment]
Would be happy to help push this through, but I barely remember any of it 😅

**Comment by bryangarza** [comment]
Gotcha, no worries, it has been quite a long time -- I think since it's been so long, we should maybe close this PR and reopen a new one if there is someone that might be able to take this forwards.

Going forwards we'll hopefully be shortening the PR feedback loop so that this doesn't happen.
Thanks!

**Pull request closed without merging**